### ISSUE TITLE
float32 refusal covers laplace only; precondition=True takes the same Hessian and fails with a misdirecting diagnosis

### ISSUE THREAD

**Issue by suchethac** [suchethac/tengri#2378]
`HESSIAN_BACKEND_SET`'s comment says the float32 refusal covers two ways of taking a Hessian. It covers one. The other fails loudly but with a misdirecting diagnosis.

Small, and not a silent failure — but the misdirection costs a reader real time, because the correct explanation already exists a few lines away in a different module.

## What the comment claims

`src/tengri/inference/mass_profile.py:657-664`:

> Backends that take a Hessian of the objective function and therefore trigger the known float32 NaN failure in the SED model's photometry Hessian […]. **The Hessian is computed via `jax.hessian` (laplace backend) or via preconditioning's `negative_hessian_metric`** (optional metric whitening for Hamiltonian samplers). Note that preconditioning is **opt-in** (`precondition=` must be truthy), so float32 refusal is scoped to the backends listed here.

Read plainly, that says: both routes take the Hessian, and the scoping to `HESSIAN_BACKEND_SET` is safe *because* the second route is opt-in.

## What the guard matches

`mass_profile.py:707` — method name only, and the set has one member:

```python
HESSIAN_BACKEND_SET = frozenset({"laplace"})
...
if method in HESSIAN_BACKEND_SET:
    dtype = jnp.result_type(float)
    if dtype == jnp.float32:
        ...
```

There is no `precondition=` term in the condition, and `grep -rn "precondition" src/tengri/inference/ | grep -iE "float32|x64|dtype"` returns only that comment. So `mcmc_nuts(precondition=True)` in float32 reaches `negative_hessian_metric` on an objective the repo has measured as all-NaN in float32 (`bench/reports/2026-09-11_profile_mass_20s.md`, Finding 8), unrefused.

## Why this is minor rather than serious

It does not fail silently. `preconditioning.py:322-331` checks the metric and raises:

```
metric is non-finite (NaN or inf), so it cannot be factorized. This is an upstream
failure, not a curvature one: the metric is built at the expansion point (normally the
MAP), so a non-finite value there (a diverged MAP, or a log-density that is NaN at that
point) propagates straight into the metric. The eigenvalue floor cannot repair it. Check
the initial point is finite, or pass precondition=False to sample without whitening.
```

That message is well-built and correct in form — #1397 clearly went to some trouble to separate non-finiteness from indefiniteness. It just points a float32 user the wrong way: it offers "a diverged MAP" and "check the initial point is finite" as the candidate causes, and the actual cause is the dtype. Meanwhile `mass_profile` already owns the exact sentence for it:

> float32 mode; the SED model's photometry Hessian is all-NaN in float32 at the converged MAP (a forward-over-reverse seam in the model), not specific to profiling (reproduces with `profile_mass=False`).

So a user hits a good error message that sends them to check a converged MAP that is fine, while the diagnosis sits one module over.

## Three ways to close it, and the choice is the point

1. **Match on the flag as well**, so `precondition=` truthy in float32 refuses with `mass_profile`'s existing reason. Most faithful to what the comment already claims, and the guard then covers both routes it names.
2. **Add the dtype to the preconditioning message** — one clause naming float32 as a candidate cause when `jnp.result_type(float)` is `float32`. Smaller, keeps the failure where it happens, and does not add a second gate to keep in sync.
3. **Correct the comment** to say the guard covers `laplace` only and that preconditioning is caught downstream by the metric check. Cheapest, and honest, but leaves the misdirecting message in place.

I would take (2), possibly with (3): the failure already raises at the right moment with the right structure, so the defect is one missing clause rather than a missing guard — and a second gate keyed on a different attribute is a thing to keep in sync, which is how the original over-claim happened.

Not proposing to implement any of them here; found while correcting the `profile_mass` guard list in the docs, which had the related error of presenting float64 as an unconditional precondition (it applies to `laplace` only, and the NaN is the SED model's, not the marginal's).

**Comment by suchethac** [comment]
Claiming (review-tengri-issues wave; the orchestrator session ceded it). Draft PR #2389 on `fix/2378-float32-hessian-refusal`: option (2)+(3) from the issue — `preconditioning.py`'s non-finite-metric and non-finite-expansion-point messages append a float32 note when `jnp.result_type(float)` is float32, the `HESSIAN_BACKEND_SET` comment says laplace-only, and one shared note constant replaces the duplicated reason string. It also fixes a reviewer finding in the same block: `resolve_profile_mass_for_method` called `disable_profile_mass` before an unconditional raise, so a Fitter reused after catching that error silently fit D instead of D−1 parameters; it now raises first, pinned by a reuse-after-failure test.
